File: server/internal/domain/board/labels_test.go
package board_test

import (
	"testing"

	"example.com/project-template/internal/domain/board"
)

// This table mirrors web/src/features/board/labels.test.ts so drift between the
// two implementations of the rule shows up as a diff on both sides.
func TestReservedLabel(t *testing.T) {
	teamLabels := []string{"Team::開發組", "Team::行政組"}
	cases := []struct {
		name     string
		label    string
		reserved bool
		reason   string
	}{
		{name: "configured team label", label: "Team::開發組", reserved: true, reason: "owned by board-directory.yml"},
		{name: "unconfigured team prefix", label: "Team::新組", reserved: true, reason: "the prefix alone is reserved"},
		{name: "lifecycle label", label: "Status::Inbox", reserved: true, reason: "the board owns lifecycle"},
		{name: "legacy workflow label", label: "To Do", reserved: true, reason: "legacy workflow"},
		{name: "legacy team label", label: "組別::開發", reserved: true, reason: "legacy team naming"},
		{name: "blank", label: "   ", reserved: true, reason: "never a valid name"},
		{name: "ordinary label", label: "Backend", reserved: false},
		{name: "namespaced but unreserved", label: "Priority::High", reserved: false},
	}
	for _, testCase := range cases {
		t.Run(testCase.name, func(t *testing.T) {
			if got := board.ReservedLabel(testCase.label, teamLabels); got != testCase.reserved {
				t.Fatalf("ReservedLabel(%q) = %v, want %v (%s)", testCase.label, got, testCase.reserved, testCase.reason)
			}
		})
	}
}

func TestDeprecatedLabel(t *testing.T) {
	for _, label := range []string{"Status::Doing", "Wating", "Doing", "組別::總召"} {
		if !board.DeprecatedLabel(label) {
			t.Fatalf("DeprecatedLabel(%q) = false, want true", label)
		}
	}
	for _, label := range []string{"Backend", "Priority::High", "Team::開發組"} {
		if board.DeprecatedLabel(label) {
			t.Fatalf("DeprecatedLabel(%q) = true, want false", label)
		}
	}
}

File: web/src/features/board/labels.test.ts
import { demoBootstrap } from "@/test/demoBootstrap";
import { describe, expect, it } from "vitest";
import { canonicalClientLabels, isDeprecatedLabel, isReservedLabel } from "./labels";

// This table mirrors server/internal/domain/board/labels_test.go so drift
// between the two implementations shows up as a diff on both sides.
const reservedCases: Array<[string, boolean, string]> = [
	["Team::開發組", true, "a configured team label"],
	["Team::新組", true, "the Team:: prefix is reserved even when unconfigured"],
	["Status::Inbox", true, "lifecycle labels are owned by the board"],
	["To Do", true, "legacy workflow label"],
	["組別::開發", true, "legacy team label"],
	["   ", true, "blank names are never valid"],
	["Backend", false, "an ordinary label"],
	["Priority::High", false, "a namespaced label that is not Team:: or Status::"]
];

describe("label rules", () => {
	it.each(reservedCases)("isReservedLabel(%s) === %s (%s)", (name, expected) => {
		expect(isReservedLabel(demoBootstrap, name)).toBe(expected);
	});

	it("treats legacy workflow and lifecycle names as deprecated", () => {
		expect(isDeprecatedLabel("Status::Doing")).toBe(true);
		expect(isDeprecatedLabel("Wating")).toBe(true);
		expect(isDeprecatedLabel("Backend")).toBe(false);
	});

	it("keeps general labels and swaps in exactly the active team's label", () => {
		const team = demoBootstrap.teams[0]!;
		const other = demoBootstrap.teams[1]!;
		const result = canonicalClientLabels(demoBootstrap, ["Backend", other.gitLabLabel, "Status::Inbox", "To Do"], team.key);
		expect(result).toEqual(["Backend", team.gitLabLabel]);
	});

	it("drops the team label when the target team is unknown", () => {
		expect(canonicalClientLabels(demoBootstrap, ["Backend", "Team::開發組"], "not-a-team")).toEqual(["Backend"]);
	});
});
